### PR TITLE
Adds multiple accounts support

### PR DESCRIPTION
**Pull request by mvietri**
References #83 

Updated to master branch and I re-wrote some parts of the manager. Still not 100% sure about this -I'm still learning-

Same mechanics as app shortcuts.

- Add multiple accounts following the login flow
- Switch or delete account 
- Logout to userless mode
- Re order accounts for custom order

Some images:

menu
switch
manager
delete

### CODE REVIEW

File: app/src/main/java/me/saket/dank/data/DankSqliteOpenHelper.kt [hunk 1]
```diff
@@ -26,11 +28,16 @@ class DankSqliteOpenHelper(context: Context) : SQLiteOpenHelper(context, DB_NAME
...
+
+    if (oldVersion == 2 && newVersion == 3) {
+      db.execSQL(AccountManager.QUERY_CREATE_TABLE)
     }
   }
 
   companion object {
-    private const val DB_VERSION = 2
+    private const val DB_VERSION = 3
     private const val DB_NAME = "Dank"
```
Comment (Esselans) [inline app/src/main/java/me/saket/dank/data/DankSqliteOpenHelper.kt]: I leave it as Dank. We should change it to Dawn or something like that but it will reset all configs for the users out there

**Comment by Esselans** [comment]
Well I'm leaving these here and wait for some reviews. I've been using it for a couple of days now with no major issues but I'm sure it can be done better.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerRepository.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import io.reactivex.Completable;
+import io.reactivex.Observable;
+
+import java.util.List;
+import javax.inject.Inject;
+
+@TargetApi(Build.VERSION_CODES.N_MR1)
+public class AccountManagerRepository {
+
+  private final Application appContext;
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerRepository.java]: Unused property. I suggest following suggestions in Android Studio (should be enabled by default) which catch issues similar to this one.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerRepository.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  private final Lazy<BriteDatabase> database;
+  private final Lazy<ShortcutManager> shortcutManager;
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerRepository.java]: Unused property.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerRepository.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.squareup.sqlbrite2.BriteDatabase;
+import dagger.Lazy;
+import io.reactivex.Completable;
+import io.reactivex.Observable;
+
+import java.util.List;
+import javax.inject.Inject;
+
+@TargetApi(Build.VERSION_CODES.N_MR1)
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerRepository.java]: Why is this annotation being used here?

File: app/src/main/java/me/saket/dank/ui/subreddit/SubredditActivity.java [hunk 5]
```diff
@@ -721,7 +722,8 @@ public boolean onOptionsItemSelected(MenuItem item) {
         if (userSessionRepository.get().isUserLoggedIn()) {
           showUserProfileSheet();
         } else {
-          startActivity(LoginActivity.intent(this));
+          Intent intent = new Intent(this, AccountManagerActivity.class);
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/subreddit/SubredditActivity.java]: `AccountManagerActivity` has `intent` function which you could use.

File: app/src/main/java/me/saket/dank/ui/subreddit/UserProfileSheetView.java [hunk 6]
```diff
@@ -200,32 +202,11 @@ void onClickComments() {
...
-              () -> parentSheet.collapse(),
-              error -> {
-                logoutButton.setText(R.string.login_logout);
-
-                ResolvedError resolvedError = errorResolver.get().resolve(error);
-                resolvedError.ifUnknown(() -> Timber.e(error, "Logout failure"));
-              }
-          );
-    }
+    Intent intent = new Intent(getContext(), AccountManagerActivity.class);
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/subreddit/UserProfileSheetView.java]: `AccountManagerActivity` has `intent` function which you could use.

File: app/src/main/java/me/saket/dank/ui/subreddit/UserProfileSheetView.java [hunk 7]
```diff
@@ -200,32 +202,11 @@ void onClickComments() {
...
+    this.getContext().startActivity(intent);
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/subreddit/UserProfileSheetView.java]: No need to add `this.` prefix.

File: app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java [hunk 8]
```diff
@@ -49,6 +63,24 @@ public void removeLoggedInUsername() {
     loggedInUsername.set(EMPTY);
   }
 
+  public Completable switchAccount(String username, Context ctx) {
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java]: `ctx` is unused.

File: app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java [hunk 9]
```diff
@@ -49,6 +63,24 @@ public void removeLoggedInUsername() {
...
+        Timber.d("Currently logged as this user");
+      } else {
+        accountHelper.trySwitchToUser(username);
+        loggedInUsername.set(username);
+      }
+    } catch (Exception e) {
+      Timber.e(e, "Error while switching users");
+    }
+
+    return Completable.complete();
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java]: Wrap whole body in `Completable.fromAction` instead of this call?

File: app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java [hunk 10]
```diff
@@ -49,6 +63,24 @@ public void removeLoggedInUsername() {
     loggedInUsername.set(EMPTY);
   }
 
+  public Completable switchAccount(String username, Context ctx) {
+    try {
+      if (username == null) {
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java]: You should mark username as `@Nullable` in parameter list if it really can be null. Otherwise this check is not needed.

File: app/src/main/java/me/saket/dank/ui/user/messages/InboxActivity.java [hunk 11]
```diff
@@ -86,6 +86,7 @@ public static Intent intent(Context context, InboxFolder initialFolder) {
     Intent intent = new Intent(context, InboxActivity.class);
     intent.putExtra(KEY_EXPAND_FROM_SHAPE, (Parcelable) null);
     intent.putExtra(KEY_INITIAL_FOLDER, initialFolder);
+    intent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/user/messages/InboxActivity.java]: This seems unrelated to the whole pull request. What made you add this flag?

File: app/src/main/res/values/colors.xml [hunk 12]
```diff
@@ -116,4 +116,9 @@
     <item>#FF0097</item>
     <item>#1E7145</item>
   </array>
+  
+  <!-- Account Manager -->
```
Comment (Tunous) [inline app/src/main/res/values/colors.xml]: Is it really necessary to add dedicated colors for this screen? Aren't there any already used colors that could be reused?

For example I can see that there is already `appshortcut_swipe_delete` which uses the same red color. You could rename it to something like `destructive_action_background` and use that instead. Generally I would suggest adding as few new colors as possible to make it easier to implement light theme in future.

File: app/src/main/res/layout/activity_account_manager.xml [hunk 13]
```diff
@@ -0,0 +1,98 @@
+<?xml version="1.0" encoding="utf-8"?>
+<FrameLayout android:id="@+id/account_manager_root"
+  xmlns:android="http://schemas.android.com/apk/res/android"
+  xmlns:app="http://schemas.android.com/apk/res-auto"
+  xmlns:tools="http://schemas.android.com/tools"
+  android:layout_width="match_parent"
+  android:layout_height="match_parent"
+  android:fitsSystemWindows="true">
+
+  <me.saket.dank.widgets.ViewFlipperWithDebugPreview
```
Comment (Tunous) [inline app/src/main/res/layout/activity_account_manager.xml]: Any reason why `ViewFlipperWithDebugPreview` is used here?

File: app/src/main/res/layout/activity_account_manager.xml [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        tools:layoutManager="com.google.android.flexbox.FlexboxLayoutManager"
+        tools:listitem="@layout/list_item_account" />
+
+      <me.saket.dank.widgets.AnimatedProgressBar
+        android:id="@+id/account_progressbar"
+        style="@style/DankProgressBar.Horizontal.TopAligned"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"/>
+
+        <LinearLayout
```
Comment (Tunous) [inline app/src/main/res/layout/activity_account_manager.xml]: Wrong indentation? This doesn't look to be inside of `AnimatedProgressBar`.

File: app/src/main/res/layout/list_item_account_manager_placeholder.xml [hunk 15]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Tunous) [inline app/src/main/res/layout/list_item_account_manager_placeholder.xml]: Are there any differences between this layout file and `list_item_app_shortcut_placeholder` other than the text? If not you could rename it into a more generic name and reuse it for both cases with text binding from code.

File: app/src/main/res/values/strings.xml [hunk 16]
```diff
@@ -150,7 +151,9 @@
   <string name="userprofile_karma_count">%s karma</string>
   <string name="userprofile_loading_karma">Loading karma…</string>
   <string name="userprofile_error_user_karma_load">Couldn\'t get karma</string>
+  <string name="userprofile_deleting_account"></string>
```
Comment (Tunous) [inline app/src/main/res/values/strings.xml]: Empty string?

File: app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java [hunk 17]
```diff
@@ -35,6 +44,11 @@ public UserSessionRepository(Lazy<Reddit> reddit, @Named("user_session") RxShare
 
   public void setLoggedInUsername(String username) {
     Preconditions.checkNotNull(username, "username == null");
+    // add user to repository
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java]: Call is clear enough so the comment is not needed.

File: app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java [hunk 18]
```diff
@@ -35,6 +44,11 @@ public UserSessionRepository(Lazy<Reddit> reddit, @Named("user_session") RxShare
...
+    this.userManagementRepository.get()
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/user/UserSessionRepository.java]: Unnecessary `this.` prefix.

File: app/src/main/java/me/saket/dank/ui/subreddit/UserProfileSheetView.java [hunk 19]
```diff
@@ -200,32 +202,11 @@ void onClickComments() {
   void onClickSubmissions() {
   }
 
-  @OnClick(R.id.userprofilesheet_logout)
-  void onClickLogout(TextView logoutButton) {
-    if (confirmLogoutTimer.isDisposed()) {
-      logoutButton.setText(R.string.userprofile_confirm_logout);
-      confirmLogoutTimer = Observable.timer(5, TimeUnit.SECONDS)
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/subreddit/UserProfileSheetView.java]: You should also remove the `confirmLogoutTimer` and `logoutDisposable` fields as they are no longer used.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManager.java [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+import me.saket.dank.ui.accountmanager.AccountManagerScreenUiModel;
+import me.saket.dank.ui.accountmanager.AutoValue_AccountManager;
+import me.saket.dank.utils.Cursors;
+
+@AutoValue
+public abstract class AccountManager implements AccountManagerScreenUiModel {
+
+  static final String TABLE_NAME = "Account";
+  static final String COLUMN_USERNAME = "username";
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManager.java]: This column is not clear to me.

1. The type in create table query is set to text
2. Mapper maps it to int
3. Mapper reads it as user
4. Create method uses it as rank
5. toValues method also uses it as rank
6. WHERE_USERNAME doesn't use it but uses label column

You see the inconsistencies? I guess what you want here are columns named rank and username instead. You should use the same naming everywhere.

File: app/src/main/java/me/saket/dank/data/DankSqliteOpenHelper.kt [hunk 21]
```diff
@@ -17,6 +18,7 @@ class DankSqliteOpenHelper(context: Context) : SQLiteOpenHelper(context, DB_NAME
     db.execSQL(CachedMessage.QUERY_CREATE_TABLE)
     db.execSQL(PendingSyncReply.QUERY_CREATE_TABLE)
     db.execSQL(AppShortcut.QUERY_CREATE_TABLE)
+    db.execSQL(AccountManager.QUERY_CREATE_TABLE)
   }
 
   override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/data/DankSqliteOpenHelper.kt]: The upgrade method will need some fixing. I can see that you followed the convention in the first if statement but it won't work correctly.

Let's say that the user is on the version of app where the database has version 1. Now when they update to latest version this method will be called with parameters `oldVersion=1` and `newVersion=3`. In that case by following the logic inside of if statements we can see that none of them will be executed and the upgrade won't work correctly.

In the first statement you want to keep on check on `oldVersion` only. That is, we should simply delete cached messages table when upgrading from that version.

In the second statement you want to check only on `newVersion` since you need to create the table also when upgrading from older versions. Additionally I'm not sure but perhaps we should go a bit further and check whether new version is equal to or greater than 3 to also take into account future versions.

___

Other than that it would be a good idea to add tests for the whole upgrade since that is an important operation. Of course tests for other operations on the database would be nice too but I feel like this one is the minimum.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java [hunk 22]
```diff
@@ -0,0 +1,325 @@
...
+import me.saket.dank.ui.subscriptions.SubscriptionRepository;
+import me.saket.dank.utils.ItemTouchHelperDragAndDropCallback;
+import me.saket.dank.utils.RxDiffUtil;
+import me.saket.dank.utils.itemanimators.SlideUpAlphaAnimator;
+import me.saket.dank.widgets.swipe.RecyclerSwipeListener;
+
+@DeepLink(AccountManagerActivity.DEEP_LINK)
+@RequiresApi(Build.VERSION_CODES.N_MR1)
+public class AccountManagerActivity extends DankActivity {
+  public static final String DEEP_LINK = "dank://accountManager";
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: `dawn://accountManager` perhaps?

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java [hunk 23]
```diff
@@ -0,0 +1,325 @@
...
+import me.saket.dank.ui.user.UserSessionRepository;
+import me.saket.dank.ui.accountmanager.AccountManagerAdapter.AccountManagerViewHolder;
+import me.saket.dank.ui.subscriptions.SubscriptionRepository;
+import me.saket.dank.utils.ItemTouchHelperDragAndDropCallback;
+import me.saket.dank.utils.RxDiffUtil;
+import me.saket.dank.utils.itemanimators.SlideUpAlphaAnimator;
+import me.saket.dank.widgets.swipe.RecyclerSwipeListener;
+
+@DeepLink(AccountManagerActivity.DEEP_LINK)
+@RequiresApi(Build.VERSION_CODES.N_MR1)
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: Why is that?

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java [hunk 24]
```diff
@@ -0,0 +1,325 @@
...
+      };
+
+      if (ACTION == ACTION_DELETE) {
+        timerDisposable = accountManagerRepository.get().delete(this.selectedAccount)
+            .subscribeOn(io())
+            .observeOn(mainThread())
+            .subscribe(
+                () -> {
+                  //this.userSessionRepository.get().switchAccount(null, getApplicationContext());
+                  thread.start();
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: Mixing threads and rx, that asks for too much trouble :)

If I'm not mistaken you used it to delay the execution by 2 seconds. Did you try using `.delay(2, TimeUnit.SECONDS)` instead?

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java [hunk 25]
```diff
@@ -0,0 +1,325 @@
...
+  }
+
+  private Completable queueToDelete(AccountManager account){
+    this.selectedAccount = account;
+    return this.confirmAction(ACTION_DELETE);
+  }
+
+  private Completable confirmAction(int action) {
+    if (confirmTimer.isDisposed()) {
+      ACTION = action;
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: Is this necessary? I can see that you are using `ACTION` only inside of this method. You could use the value passed to parameter directly and remove the field.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java [hunk 26]
```diff
@@ -0,0 +1,325 @@
...
+    super.finish();
+  }
+
+  @OnClick(R.id.account_manager_logout)
+  public void logout() {
+    confirmAction(ACTION_LOGOUT);
+  }
+
+  private Completable queueToDelete(AccountManager account){
+    this.selectedAccount = account;
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: Unnecessary `this.` prefix.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java [hunk 27]
```diff
@@ -0,0 +1,325 @@
...
+        // Stuff that updates the UI
+        logoutButton.setText(confirmText);
+        logoutButton.setVisibility(View.VISIBLE);
+      });
+
+
+      confirmTimer = Observable.timer(5, TimeUnit.SECONDS)
+          .compose(applySchedulers())
+          .subscribe(o -> {
+            runOnUiThread(() -> {
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: The call to `applySchedulres` above sets observation to happen on the main thread which makes this call redundant.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java [hunk 28]
```diff
@@ -0,0 +1,325 @@
...
+  }
+
+  private void setupUserList() {
+    SlideUpAlphaAnimator animator = SlideUpAlphaAnimator.create();
+    animator.setSupportsChangeAnimations(false);
+    accountsRecyclerView.setItemAnimator(animator);
+    accountsRecyclerView.setLayoutManager(new LinearLayoutManager(this));
+    accountsRecyclerView.setAdapter(accountManagerAdapter.get());
+
+    //fullscreenProgressView.setVisibility(View.VISIBLE);
```
Comment (Tunous) [inline app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: If there is any commented-out code it should be removed.

File: app/src/main/java/me/saket/dank/data/DankSqliteOpenHelper.kt (hunk 1 above)
Comment (Tunous) [reply app/src/main/java/me/saket/dank/data/DankSqliteOpenHelper.kt]: There is no need for this change. Similarly to package name we would only lose stuff by switching to Dawn so it should be left as is.

**Comment by Esselans** [comment]
Hey @Tunous Thanks for your feedback and review. I'll update the PR (I didn't have the time yet sadly).

File: app/src/main/java/me/saket/dank/data/InboxRepository.java [hunk 29]
```diff
@@ -258,6 +258,16 @@ private Completable removeAllMessages(InboxFolder folder) {
     });
   }
 
+  @CheckResult
+  public Completable clearMessages() {
+    return Completable.fromAction(() -> {
+      try (BriteDatabase.Transaction transaction = briteDatabase.newTransaction()) {
+        briteDatabase.delete(CachedMessage.TABLE_NAME, "");
```
[the remark this answers is not in the document]
Comment (Esselans) [reply app/src/main/java/me/saket/dank/data/InboxRepository.java]: Ah OK. Anyways I wasn't 100% sure about deleting the cache table but switching to other account would end in merging PM from all your accounts. It's seems to be more a hack than a fix to me.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java [hunk 30]
```diff
@@ -0,0 +1,325 @@
...
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_account_manager);
+    ButterKnife.bind(this);
+  }
+
+  @Override
+  protected void onPostCreate(@Nullable Bundle savedState) {
+    super.onPostCreate(savedState);
+
+    setupUserList();
```
[the remark this answers is not in the document]
Comment (Esselans) [reply app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: Hmm I have to check where I copy that from. Maybe it's wrong on both side.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java (hunk 24 above)
Comment (Esselans) [reply app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: Yeah. Rx and me... we don't get along very well haha. When the PR is updated I'll let someone take care of this issue.

File: app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java (hunk 25 above)
Comment (Esselans) [reply app/src/main/java/me/saket/dank/ui/accountmanager/AccountManagerActivity.java]: Will check.

**Comment by Tunous** [comment]
Superseded by #329 which continued on work from this pull request. @mvietri Thank you for the original work.
